NixOS/nixpkgs
# musl-cross, native-musl, cross fixes, oh my!

## Motivation for this change

Support cross-compiling for musl, support musl-native, generalize a bit beyond glibc, many cross fixes.

A few fixes are "borrowed" from @bgamari's big PR (#30882).

## PR History

To help folks track changes made across updates to this PR, I'm tagging each revision to help diff and whatnot:

* [v1 - 2018-02-05](https://github.com/dtzWill/nixpkgs/releases/tag/musl-pr-1)
  * Initial submission.
* [v2 - 2018-02-06](https://github.com/dtzWill/nixpkgs/releases/tag/musl-pr-2)
  * Drop musl.nix, musl-cross.nix (commits introducting and modifying)
  * Drop 305a63407ed352bda7b13f432be2c79df2bde4be (gcc6 yesFuncs/noFuncs):
    * doesn't seem needed, and as a bonus wasn't used for stdenv.cc anyway so no mass rebuild
  * busybox: don't force gcc
  * libcap: don't force gcc, also cleanup to use depsBuildBuild and existing env vars
  * musl: don't redundantly add cross prefix
  * Drop stdenv.is{Glibc,Musl}, grab from hostPlatform
  * Drop openjdk commit/revert, will revisit later
  * pam: depsBuildBuild
* [v3 - 2018-02-08](https://github.com/dtzWill/nixpkgs/releases/tag/musl-pr-3)
  * rebase onto latest master
  * resolve squid4 conflict by dropping commit: fix no longer needed, can enable parallel building in separate PR
  * re-enable LLVM manpage building on musl, only disable on cross (for now).
* [v4 - 2018-02-08](https://github.com/dtzWill/nixpkgs/releases/tag/musl-pr-4)
  * drop 'libc = "glibc"' from system examples
  * rebase onto staging, since that's what we're targeting.  Whoops.
* [v5 - 2018-02-09](https://github.com/dtzWill/nixpkgs/releases/tag/musl-pr-5)
  * specify commit hash in gcc5 musl patch URL
  * resolve conflict with nixUnstable, rebase onto latest staging
  * fix typo in "examples" for aarch64-musl introduced in v4
  * tweak llvm-{4,5} use of python to (hopefully) fix darwin eval, allow either python2 or python3
  * fix go eval problems by replacing attempts to use `stdenv.glibc`
* [v6 - 2018-02-11](https://github.com/dtzWill/nixpkgs/releases/tag/musl-pr-6)
  * resolve conflicts, rebase onto latest staging
  * groundwork for musl32, although no bootstrap here and cross doesn't work yet
  * restore "stdenv.glibc", too messy to remove as part of this and can be done later.
* [v7 - 2018-02-11](https://github.com/dtzWill/nixpkgs/releases/tag/musl-pr-7)
  * refactor lib changes into #34861, rebase onto that and latest staging
  * drop nearly all LLVM/clang fixes from this, will tackle in follow-up
  * lots of cleanup, esp. re:isGlibc/isMusl and breaking apart / merging commits (**~50 fewer commits!**)
* [v8 - 2018-02-11](https://github.com/dtzWill/nixpkgs/releases/tag/musl-pr-8)
  * fetch patches instead of local copies: net-snmp, apparmor, audit
  * don't use IFD for systemd patch list and instead generate postPatch command
* [v9 - 2018-02-12](https://github.com/dtzWill/nixpkgs/releases/tag/musl-pr-9)
  * drop unused python patch
  * support musl-native bootstrapping on aarch64, armv6l (in addition to x86_64).  Not tested yet :innocent:.
* [v10 - 2018-02-13](https://github.com/dtzWill/nixpkgs/releases/tag/musl-pr-10)
  * rebase onto latest staging, resolve systemd 237 conflicts by dropping patches used to get it building.  Will revisit later due to effort involved and amount of patches required.
  * use 64bit busybox for bootstrap on x86_64 to avoid problems like those reported in WSL.
  * actually restore "stdenv.glibc" (for now), previous commit did not do this for final stdenv
  * tested bootstrap on aarch64 (hooray and thank you @grahamc and https://github.com/nix-community/aarch64-build-box), it's at least not completely broken.
  * includes XZ fix (#34916) since without it stdenv contains invalid referencces

## TODO

* [x] use fetchpatch instead of introducing local files, make a pass checking for this

### Minor / Misc / Maybe
* [ ] cc-wrapper: When building gcc6 (wrapped) in cross stdenv, somehow binutils from host get onto PATH (?) causing the wrong `mkdir` to be used.  This doesn't occur when building the cross `stdenv.cc`, FWIW. Example error:
    ```
    /nix/store/k1r93q2cq3r7x8kx663w8lf0qaid7iiv-stdenv/setup: line 1224: /nix/store/ldsqyff47p1v0yqjp4a3kzxcqvv5fmwr-coreutils-8.29-armv6l-unknown-linux-musleabihf/bin/mkdir: cannot execute binary file: Exec format error
    ```
* [ ] refactor duplicated code compute path to `getent` utility

### Quickstart with Binary Cache

To enjoy musl goodness right away, consider taking advantage of the ALLVM project's binary cache!

To do so, add the following to your configuration.nix:

```nix
{
    nix.binaryCaches = [
      "https://cache.nixos.org/"
      "https://cache.allvm.org/"
    ];

    nix.binaryCachePublicKeys = [
      "gravity.cs.illinois.edu-1:yymmNS/WMf0iTj2NnD0nrVV8cBOXM9ivAkEdO1Lro3U="
    ];
}
```

(Please add it after the nixos one O:))

Equivalent of "small" jobset is evaluated and cached for a few build configurations,
but certainly not everything works yet.  This is just a (hopefully solid) start!

----------------

### Credits and Teasers: ALLVM

This PR (and many of the related fixes submitted recently) are based on work done as part of the [ALLVM project](http://allvm.org), many thanks to our funding :+1:.

We look forward to sharing more with everyone soon! Stay tuned :).
